From fcb468e767c0dc02876c9a82a16037d3ec076d0d Mon Sep 17 00:00:00 2001
From: Andreas Scherer <andreas_github@freenet.de>
Date: Sun, 8 Jan 2017 15:13:01 +0100
Subject: [PATCH 29/43] DRY up the Makefile.

The rules for mmix-pipe and mmix-sim are similar. With the
.SECONDEXPANSION feature of GNU make, this can be made clear.

And we can replace 'abstime' by 'perl/date'. No need to work with
the temporary file 'abstime.h' at all. Note: 'perl' is used because
'sed' on Linux and BSD (macos) have incompatible invocation options.

This brings ideas from branches 'make-mmix-sim.o' and 'secondexpansion'
into production.
---
 Makefile | 21 +++++++++------------
 1 file changed, 9 insertions(+), 12 deletions(-)

diff --git a/Makefile b/Makefile
index b46d129..f6afd12 100644
--- a/Makefile
+++ b/Makefile
@@ -73,27 +73,24 @@ all:    mmixal mmix mmotype mmmix
 clean:
 	rm -f *~ *.o *.c *.h *.tex *.log *.dvi *.toc *.idx *.scn *.ps core
 
-mmix-pipe.o: mmix-pipe.c abstime
-	./abstime > abstime.h
-	$(CC) $(CFLAGS) -c mmix-pipe.c
-	rm abstime.h
+.SECONDEXPANSION:
+mmix-pipe.o mmix-sim.o: $$(subst .o,.c,$$@)
+	perl -pe "s/(#define ABSTIME)/\1 `date +%s`/" -i $<
+	$(CC) $(CFLAGS) -c $<
 
 mmix-config.o: mmix-pipe.o
 
 mmmix:  mmix-arith.o mmix-pipe.o mmix-config.o mmix-mem.o mmix-io.o mmmix.c
-	$(CC) $(CFLAGS) mmmix.c \
-	  mmix-arith.o mmix-pipe.o mmix-config.o mmix-mem.o mmix-io.o -o mmmix
+	$(CC) $(CFLAGS) $^ -o $@
 
 mmixal: mmix-arith.o mmixal.c
-	$(CC) $(CFLAGS) mmixal.c mmix-arith.o -o mmixal
+	$(CC) $(CFLAGS) $^ -o $@
 
-mmix:   mmix-arith.o mmix-io.o mmix-sim.c abstime
-	./abstime > abstime.h
-	$(CC) $(CFLAGS) mmix-sim.c mmix-arith.o mmix-io.o -o mmix
-	rm abstime.h
+mmix:   mmix-arith.o mmix-io.o mmix-sim.o
+	$(CC) $(CFLAGS) $^ -o $@
 
 mmotype: mmotype.c
-	$(CC) $(CFLAGS) mmotype.c -o mmotype
+	$(CC) $(CFLAGS) $^ -o $@
 
 tarfile: $(ALL)
 	tar cvf /tmp/mmix.tar $(ALL)
-- 
2.19.0

